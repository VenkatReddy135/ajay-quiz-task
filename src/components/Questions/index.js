import React, { useState, useEffect, memo } from "react";
import quizQuestions from "../../Api/questions";
import Quiz from "../../sharedComponent/Quiz/Quiz";
import Result from "../../sharedComponent/Result";
import Timer from "../../sharedComponent/Timer/Timer";
import {
  callSelectedAnswers,
  setQuestion,
} from "../../sharedComponent/HelperFunction";

const Questions = () => {
  const initialState = {
    questionId: 1,
    question: "",
    answerOptions: [],
    allQuestions: [],
    counter: 0,
    selectedAnswers: {},
    result: "",
  };
  const [input, setInput] = useState(initialState);

  useEffect(() => {
    setInput((prevState) => ({
      ...prevState,
      question: quizQuestions[0].question,
      answerOptions: quizQuestions[0].answers,
      allQuestions: quizQuestions,
    }));
  }, []);

  const handleAnswerSelected = (e) => {
    let answerObject = callSelectedAnswers(input, e);
    setInput((prevState) => ({
      ...prevState,
      selectedAnswers: answerObject,
    }));
  };

  const setNextQuestion = () => {
    const next = "next";
    let nextQuestion = setQuestion(input, next);
    setInput((prevState) => ({
      ...prevState,
      ...nextQuestion,
    }));
  };

  const setPreviousQuestion = () => {
    const previous = "previous";
    let previousQuestion = setQuestion(input, previous);
    setInput((prevState) => ({
      ...prevState,
      ...previousQuestion,
    }));
  };

  const ResultDetails = memo(() => {
    return (
      <Result quizResult={input.allQuestions} answers={input.selectedAnswers} />
    );
  });

  const viewResults = () => {
    setInput((prevState) => ({
      ...prevState,
      result: true,
    }));
  };

  const QuizDetails = () => {
    return (
      <Quiz
        viewResults={viewResults}
        setNextQuestion={setNextQuestion}
        counter={input.counter}
        setPreviousQuestion={setPreviousQuestion}
        selectedAnswer={input.selectedAnswers[input.counter]}
        answerOptions={input.answerOptions}
        questionId={input.questionId}
        question={input.question}
        questionTotal={quizQuestions.length}
        onAnswerSelected={handleAnswerSelected}
      />
    );
  };

  return input.result ? (
    <ResultDetails />
  ) : (
    <>
      <Timer viewResults={viewResults} /> <QuizDetails />
    </>
  );
};

export default React.memo(Questions);
