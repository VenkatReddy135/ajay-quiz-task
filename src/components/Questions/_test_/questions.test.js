/* eslint-disable testing-library/no-node-access */
import React from "react";
import { cleanup } from "@testing-library/react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import Questions from "../index";
import Enzyme from "enzyme";
import Adapter from "enzyme-adapter-react-17-updated";
Enzyme.configure({ adapter: new Adapter() });

describe("Questions component", () => {
  const renderedData = renderer.create(<Questions />).toJSON();

  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<Questions />, div);
  });

  it("renders a snapshot", () => {
    expect(renderedData).toMatchSnapshot();
  });

  afterEach(() => {
    cleanup();
  });
});
