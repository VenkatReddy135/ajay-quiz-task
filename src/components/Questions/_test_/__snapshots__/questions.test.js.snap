// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Questions component renders a snapshot 1`] = `
Array [
  <div
    className="container"
    style={
      Object {
        "marginBottom": "1em",
        "marginTop": "2em",
      }
    }
  >
    <button
      className="btn btn-info"
      type="button"
      value=""
    >
      3:20
    </button>
  </div>,
  " ",
  <div
    className="container"
  >
    <div
      className="questionCount"
    >
      Question 
      <span>
        1
      </span>
       of 
      <span>
        6
      </span>
    </div>
    <div>
      <p
        className="fs-2"
      >
        
      </p>
    </div>
    <ul
      className="answerOptions"
    />
    <div
      className="bottom-footer"
    >
      <div />
      <button
        className="btn btn-warning buttonMargin"
        onClick={[Function]}
        type="button"
        value=""
      >
        Next
      </button>
    </div>
    <div
      style={
        Object {
          "display": "flex",
          "flexDirection": "row-reverse",
        }
      }
    />
  </div>,
]
`;
